py/objtype.c

```diff
@@ -1014,6 +1014,21 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
             dest[0] = MP_OBJ_NEW_QSTR(self->name);
             return;
         }
+        if (attr == MP_QSTR___bases__) {
+            if (self == &mp_type_object) {
+                dest[0] = mp_obj_new_tuple(0, NULL);
```

Comment:
I think it'd be simpler to use `dest[0] = mp_const_empty_tuple;`

Reply:
I thought you might mention that. I chose calling the API over assigning to the constant because I told myself it was better to call the API than to use constants.

Though it is part of the core and we don't really have an API, now that I'm typing out my reasoning.

Reply:
Couldn't think of any good reason to defend my initial assumption. Pushing the change.

Reply:
The call to `mp_obj_new_tuple(0, NULL)` will just return `mp_const_empty_tuple` so it's efficient in the sense of not allocating RAM.

But I'd encourage use of `mp_const_empty_tuple` because it leads to smaller and faster code.  This constant (along with the others) are really part of the C API (if there was a formal API).

Reply:
> because it leads to smaller and faster code

I was wondering about that, since it doesn't take much static analysis to conclude that if the `size_t n` is equal to the constant 0, then the result is equal to the constant `mp_const_empty_tuple` would GCC not magic that all away?

I'm completely happy with using `mp_const_empty_tuple` but I am wondering how optimized compilers are.

---

I think this PR is ready. Thanks for your help reviewing this.

Reply:
Not sure if I've implemented a simliar enough case here, but if its of interest to you:

http://coliru.stacked-crooked.com/a/ac3332e785890500

```
movq	$_ZL23example_struct_zero_obj, -8(%rsp)
```

Reply:
FWIW, @jimmo gave me a C leason on Slack and my hypothesis is completely wrong. Even in what is a purely functional (and statically determinate case). GCC would not collapse the function call away :smile:

Reply:
It may when using LTO (link time optimisation) but it'd need to be very smart.

Reply:
I tried with `-flto` and it still made a call. We're just not there yet.